plugins/usertiming.js

```diff
@@ -0,0 +1,93 @@
+/**
+\file usertiming.js
+Plugin to collect metrics from the W3C User Timing API.
+For more information about User Timing,
+see: http://www.w3.org/TR/user-timing/
+*/
+
+(function() {
+
+	BOOMR = BOOMR || {};
+	BOOMR.plugins = BOOMR.plugins || {};
+
+	/**
+	 * @returns String compressed user timing data that occurred since the last call
+	 */
+	function getUserTiming() {
+		var now = BOOMR.now();
+
+		map = UserTimingCompression.getCompressedUserTiming(impl.options);
+		res = UserTimingCompression.compressForUri(map);
+		impl.options["from"] = now;
+
+		return res;
+	}
+
+	var impl = {
+		complete: false,
+		initialized: false,
+		supported: false,
+		options: {"from": 0},
+
+		done: function() {
+			var r;
+
+			if (this.complete) {
+				return;
+			}
+
+			BOOMR.removeVar("usertiming");
+			r = getUserTiming();
+			if (r) {
+				BOOMR.info("Client supports User Timing API", "usertiming");
+				BOOMR.addVar({
+					"usertiming": JSON.stringify(r)
+				});
+			}
+
+			this.complete = true;
+			BOOMR.sendBeacon();
+		},
+
+		clearMetrics: function(vars) {
+			if (vars.hasOwnProperty("usertiming")) {
+				BOOMR.removeVar("usertiming");
+			}
+			this.complete = false;
+		}
+	};
+
+	BOOMR.plugins.UserTiming = {
+		init: function(config) {
+			var p = BOOMR.window.performance;
+
+			BOOMR.utils.pluginConfig(impl, config, "UserTiming");
+
+			if (impl.initialized) {
+				return this;
+			}
+
+			if (p && typeof p.getEntriesByType === "function") {
+				BOOMR.subscribe("page_ready", impl.done, null, impl);
```

Comment:
I think rather than subscribe to `page_ready` and `xhr_load`, it might be better to just subscribe to `before_beacon` and add parameters at that time.

This way we also capture data if beacons were fired some other way, and we don't stomp over ourselves if, for example, `page_ready` were called, but a beacon wasn't sent because another plugin was not ready.